### DISCUSSION TITLE
Should we have the error message "Expected null, received empty string"

### DISCUSSION THREAD

**Discussion by rsinohara** [colinhacks/zod#633] [Ideas]
I often see mismatched types mixing null and empty strings. We could make the error message differentiate between empty strings in those cases, since it can point to a clearer schema issue: a fix to the schema should be `literal("")`  instead of `string()`.

**Comment by rsinohara** [comment]
Context: I am just fixing my schemas after an API update that changed null to empty strings. The values are not strings, which makes matters worse, the null value was just replaced by "".

I was adding `.or(z.string())` until I realized the situation.

**Reply by StefanTerdell** [reply]
"" is still technically a string, though. Maybe you're looking for z.string().nonempty()?

**Reply by rsinohara** [reply]
In my case I just wanted z.literal('') really, because the value was always null-ish. I just wish the message told me that instead of null, it received an empty string. The current error says string, which let me to thing it was an actual non empty string.

The error right now is not wrong, but in my case just a little more detail would have helped.

**Reply by scotttrinh** [reply]
This a bit of a slippery slope, though: "Expected null, received 0" or "Expected null, received false", etc. I think "Expected null, received string" is correct when you give a `z.null()` a string, regardless of if it's falsy or not. Using the common usage of "nullish" to mean `null | undefined`, `""` isn't really nullish, it's falsy, so I would expect that if we special cased one falsy value it would be weird to not special case the others.

**Reply by rsinohara** [reply]
That makes sense.
What if we simply  the actual value to the error `Expected null, received String ""`. That would work for any type.

**Reply by scotttrinh** [reply]
Yeah, that's a good suggestion! We've gone around a bit in the past about exposing the input values in error messages, since they may contain potentially sensitive information.

**Reply by StefanTerdell** [reply]
> We've gone around a bit in the past about exposing the input values in error messages, since they may contain potentially sensitive information.

I agree with this concern fwiw. At the very least such a change should probably come with a major version, unless disabled by default through some kind of option pattern. Perhaps through a functional overload to custom error messages? Like so: 
``` typescript
z.number().min(1, { message: (value) => `Value should be atleast 1 but was ${value}` })
```
Or perhaps with a flag next to the `message?` prop
``` typescript
z.number().min(1, { logInvalidValue: true })
```
